pkg/smi/health.go

```diff
@@ -0,0 +1,64 @@
+package smi
+
+import (
+	smiAccess "github.com/servicemeshinterface/smi-sdk-go/pkg/apis/access/v1alpha3"
+	smiSpecs "github.com/servicemeshinterface/smi-sdk-go/pkg/apis/specs/v1alpha4"
+	smiSplit "github.com/servicemeshinterface/smi-sdk-go/pkg/apis/split/v1alpha2"
+	extensionsClientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/client-go/rest"
+
+	"github.com/openservicemesh/osm/pkg/version"
+)
+
+// HealthChecker has SMI kube config to access SMI CRDS
+type HealthChecker struct {
+	SMIKubeConfig *rest.Config
+}
+
+// Liveness is the Kubernetes liveness probe handler.
+func (smi HealthChecker) Liveness() bool {
+	return true
+}
+
+// Readiness is the Kubernetes readiness probe handler.
+func (smi HealthChecker) Readiness() bool {
+	return checkSMICrdsExist(smi.SMIKubeConfig)
+}
+
+// GetID returns the ID of the probe
+func (smi HealthChecker) GetID() string {
+	return "SMI"
+}
+
+func checkSMICrdsExist(SMIKubeConfig *rest.Config) bool {
+	client := extensionsClientset.NewForConfigOrDie(SMIKubeConfig)
+
+	// The key is the API Resource.Kind
+	// The value is the SMI CRD Type
+	reqVersions := map[string]string{
+		"HTTPRouteGroup": "specs",
+		"TCPRoute":       "specs",
+		"TrafficSplit":   "split",
+		"TrafficTarget":  "access",
+	}
+	var candidateVersions = []string{smiAccess.SchemeGroupVersion.String(), smiSpecs.SchemeGroupVersion.String(), smiSplit.SchemeGroupVersion.String()}
+
+	for _, groupVersion := range candidateVersions {
+		list, err := client.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return false
+		}
```

Comment:
Let's at least log this error so we can tell if this probe is failing because the CRDs are really missing or if there's some other error like connectivity with the API server.